File: packages/ts/src/zyins/branding.ts
/**
 * Tier 3 branding operations — GET `/v1/branding`.
 *
 * Branding is per-license-order whitelabel configuration: agency name, logo
 * URL, colors, and product restrictions. Identity comes from the
 * License-HMAC auth headers; the request carries no body credentials. See
 * `docs/design/cases-email-branding-surface.md` for the #149 auth elevation
 * — when session credentials replace License-HMAC, this SDK surface is
 * unaffected (no method args change).
 *
 * Returns a zero-value `BrandingDetail` when no row exists; the server
 * deliberately does NOT 404 for missing branding rows.
 */

import { type AuthContext } from './auth';
import { type Transport } from './transport';
import { fromHttpResponse } from './errors';
import { boolField, isRecord, parseJsonResponse, stringField, unwrapEnvelope } from './response';
import { buildLicenseHMACHeaders } from '../core';
import { type Clock, systemClock } from '../core';

const BRANDING_PATH = '/v1/branding';

/** Whitelabel detail returned by `branding.lookup`. */
export interface BrandingDetail {
  imoName: string;
  imoLogo: string;
  navColor: string;
  mainColor: string;
  buttonColor: string;
  activeButtonColor: string;
  bgColor: string;
  headerTextColor: string;
  hideAffiliateLeads: boolean;
  preventProductSelection: boolean;
  defaultSettings: string;
}

/** Per-call context. */
export interface BrandingContext {
  baseUrl: string;
  auth: AuthContext;
  transport: Transport;
  clock: Clock;
}

/** Fetch the whitelabel branding for the caller's license. */
export async function lookup(ctx: BrandingContext): Promise<BrandingDetail> {
  const headers = await buildLicenseHMACHeaders(
    ctx.auth.licenseKey,
    ctx.auth.orderId,
    ctx.auth.email,
    'GET',
    BRANDING_PATH,
    '',
    ctx.auth.deviceId,
    ctx.clock ?? systemClock,
  );
  const response = await ctx.transport({
    url: `${ctx.baseUrl}${BRANDING_PATH}`,
    method: 'GET',
    headers: { ...headers, Accept: 'application/json' },
    body: '',
  });
  if (response.status >= 200 && response.status < 300) {
    return parseBrandingResponse(response.body);
  }
  throw fromHttpResponse(response.status, response.body);
}

function parseBrandingResponse(body: string): BrandingDetail {
  if (!body) {
    return zeroBrandingDetail();
  }
  const parsed = parseJsonResponse(body, 'branding.lookup');
  const unwrapped = unwrapEnvelope(parsed);
  const root = isRecord(unwrapped) ? unwrapped : {};
  return {
    imoName: stringField(root, 'imo_name'),
    imoLogo: stringField(root, 'imo_logo'),
    navColor: stringField(root, 'nav_color'),
    mainColor: stringField(root, 'main_color'),
    buttonColor: stringField(root, 'button_color'),
    activeButtonColor: stringField(root, 'active_button_color'),
    bgColor: stringField(root, 'bg_color'),
    headerTextColor: stringField(root, 'header_text_color'),
    hideAffiliateLeads: boolField(root, 'hide_affiliate_leads'),
    preventProductSelection: boolField(root, 'prevent_product_selection'),
    defaultSettings: stringField(root, 'default_settings'),
  };
}

function zeroBrandingDetail(): BrandingDetail {
  return {
    imoName: '',
    imoLogo: '',
    navColor: '',
    mainColor: '',
    buttonColor: '',
    activeButtonColor: '',
    bgColor: '',
    headerTextColor: '',
    hideAffiliateLeads: false,
    preventProductSelection: false,
    defaultSettings: '',
  };
}

File: packages/python/src/sah_sdk/zyins/branding.py
"""Branding sub-client — ``GET /v1/branding``.

Branding is per-license-order whitelabel configuration: agency name,
logo URL, colors, and product restrictions. Identity is derived from
the configured auth context — no body fields required.

The server deliberately does NOT 404 when a row is missing; it returns
a zero-value :class:`BrandingDetail`. See
``docs/design/cases-email-branding-surface.md`` for the #149 auth
elevation context — when session-credentials replace License-HMAC,
this SDK surface is unaffected.
"""

from __future__ import annotations

import json
from typing import Any

from pydantic import BaseModel, ConfigDict


class BrandingDetail(BaseModel):
    """Whitelabel detail returned by :meth:`BrandingSubClient.lookup`."""

    model_config = ConfigDict(extra="ignore", frozen=True)

    imo_name: str = ""
    imo_logo: str = ""
    nav_color: str = ""
    main_color: str = ""
    button_color: str = ""
    active_button_color: str = ""
    bg_color: str = ""
    header_text_color: str = ""
    hide_affiliate_leads: bool = False
    prevent_product_selection: bool = False
    default_settings: str = ""


def _unwrap_envelope(raw: str, *, context: str) -> dict[str, Any]:
    if not raw:
        raise ValueError(f"{context}: response body was empty")
    parsed = json.loads(raw)
    if not isinstance(parsed, dict):
        raise ValueError(f"{context}: response body was not a JSON object")
    data = parsed.get("data")
    if isinstance(data, dict):
        return data
    return parsed


def _coerce_bool(value: Any) -> bool:
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        return value in ("true", "1")
    return False


def parse_lookup_response(raw: str) -> BrandingDetail:
    """Parse a branding response, tolerating both bare and ADR-012 envelopes."""
    root = _unwrap_envelope(raw, context="branding.lookup")
    return BrandingDetail(
        imo_name=str(root.get("imo_name") or ""),
        imo_logo=str(root.get("imo_logo") or ""),
        nav_color=str(root.get("nav_color") or ""),
        main_color=str(root.get("main_color") or ""),
        button_color=str(root.get("button_color") or ""),
        active_button_color=str(root.get("active_button_color") or ""),
        bg_color=str(root.get("bg_color") or ""),
        header_text_color=str(root.get("header_text_color") or ""),
        hide_affiliate_leads=_coerce_bool(root.get("hide_affiliate_leads")),
        prevent_product_selection=_coerce_bool(root.get("prevent_product_selection")),
        default_settings=str(root.get("default_settings") or ""),
    )
